### ISSUE TITLE
Support for launchtemplate with both Ipv6AddressCount and Ipv6PrefixCount

### ISSUE THREAD

**Issue by anguslees** [aws/karpenter#2187]
## Version

**Karpenter**: v0.13.2

**Kubernetes**: v1.22.10-eks-84b4fe6

## Expected Behavior

karpenter should be able to use an AWS LaunchTemplate that attaches both an IPv6 address and an IPv6 prefix to an interface (ie: a NetworkInterface with both Ipv6AddressCount=1 _and_ Ipv6PrefixCount=1)

Creating an instance with this LaunchTemplate manually (using MNG/ASG or RunInstances, without fleet), works just fine.  Instances are created with both a single IPv6 address, and a /80 IPv6 prefix attached to primary interface at Instance creation.

(This combination is particularly interesting in EKS, since it gives an address for the node, and a prefix for pods, and means vpc-cni has nothing to do at runtime.  vpc-cni can be run without any mutation IAM perms - indeed, vpc-cni can even be replaced with much simpler alternatives like `host-local`+`ptp`.)

## Actual Behavior

karpenter fails with
```
Provisioning failed, launching node, creating cloud provider instance, with fleet error(s), InvalidParameterValue: You can and can only specify one of ipv6-addresses or ipv6-addresses-count  ipv6-prefix and ipv6-prefix-count.; MaxSpotInstanceCountExceeded: Max spot instance count exceeded; UnsupportedOperation: IPv6 is not supported for the instance type 'm3.medium'.
```

I suspect this is actually a bug in fleet's validation of this request, since I note CloudFormation has the same validation bug.

## Steps to Reproduce the Problem

1. Install terraform.
2. Copy below `main.tf` into a new directory
3. `terraform init`
4. `terraform apply`

Note karpenter runs (in fargate) but node creation fails.  See karpenter pod logs for above error message.
Comment out `ipv6_prefix_count = 1` from launch template to see node creation succeed, but without a prefix attached at creation time.

## Resource Specs and Logs

<details>
<summary>main.tf</summary>

(Edit: updated to restrict to nitro instance types)

```
terraform {
  required_version = ">= 1.1.0"
  required_providers {
    aws = {
      source = "hashicorp/aws"
      # Need >= 4.6 because launchtemplate ipv6_prefix_count
      # avoid 4.6.0 because https://github.com/hashicorp/terraform-provider-aws/issues/23766
      version = "~> 4.6, != 4.6.0"
    }

    helm = {
      source = "hashicorp/helm"
      version = "~> 2.5"
    }

    kubectl = {
      source = "gavinbunney/kubectl"
      version = "~> 1.14"
    }

    cloudinit = {
      source = "hashicorp/cloudinit"
      version = "~> 2.2"
    }
  }
}

provider "aws" {
  region = local.region
  
  default_tags {
    tags = {
      testcase = local.name
    }
  }
}

provider "kubernetes" {
  host = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
  exec {
    api_version = "client.authentication.k8s.io/v1beta1"
    command = "aws"
    args = ["eks", "get-token", "--cluster-name", module.eks.cluster_id, "--region", local.region]
  }
}

provider "helm" {
  kubernetes {
    host = module.eks.cluster_endpoint
    cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
    exec {
      api_version = "client.authentication.k8s.io/v1beta1"
      command = "aws"
      args = ["eks", "get-token", "--cluster-name", module.eks.cluster_id, "--region", local.region]
    }
  }
}

provider "kubectl" {
  apply_retry_count = 5
  host = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
  load_config_file = false
  exec {
    api_version = "client.authentication.k8s.io/v1beta1"
    command = "aws"
    args = ["eks", "get-token", "--cluster-name", module.eks.cluster_id, "--region", local.region]
  }
}

locals {
  name = "karpenter-v6bug"
  region = "ap-southeast-2"

  cidr_v4 = "10.0.0.0/16"
}

module "eks" {
  source = "terraform-aws-modules/eks/aws"
  version = "~> 18"

  cluster_name = local.name
  cluster_endpoint_private_access = true
  cluster_endpoint_public_access = true

  cluster_ip_family = "ipv6"
  create_cni_ipv6_iam_policy = true

  cluster_addons = {
    coredns = {
      resolve_conflicts = "OVERWRITE"
    }
    kube-proxy = {
      resolve_conflicts = "OVERWRITE"
    }
    vpc-cni = {
      resolve_conflicts = "OVERWRITE"
      service_account_role_arn = module.vpc_cni_irsa.iam_role_arn
    }
  }

  enable_irsa = true
  
  vpc_id = module.vpc.vpc_id
  subnet_ids = module.vpc.private_subnets
  control_plane_subnet_ids = module.vpc.intra_subnets

  cluster_encryption_config = [{
    provider_key_arn = aws_kms_key.eks.arn
    resources = ["secrets"]
  }]
  
  cluster_security_group_additional_rules = {
    egress_nodes_ephemeral_ports_tcp = {
      description                = "To node 1025-65535"
      protocol                   = "tcp"
      from_port                  = 1025
      to_port                    = 65535
      type                       = "egress"
      source_node_security_group = true
    }
  }

  fargate_profiles = {
    karpenter = {
      name = "karpenter"
      selectors = [{namespace = "karpenter"}]
    }
  }
  
  node_security_group_ntp_ipv4_cidr_block = ["169.254.169.123/32"] # default 0/0
  node_security_group_ntp_ipv6_cidr_block = ["fd00:ec2::123/128"]  # default ::/0

  node_security_group_tags = {
    # Must match karpenter_irsa.karpenter_tag_key in generated IAM policy
    # See also https://github.com/aws/karpenter/issues/1919
    "karpenter.sh/managed-by" = local.name
  }
  
  node_security_group_additional_rules = {
    ingress_self_all = {
      description = "Node to node all ports/protocols"
      protocol    = "-1"
      from_port   = 0
      to_port     = 0
      type        = "ingress"
      self        = true
    }

    egress_all = {
      description      = "Node all egress"
      protocol         = "-1"
      from_port        = 0
      to_port          = 0
      type             = "egress"
      cidr_blocks      = ["0.0.0.0/0"]
      ipv6_cidr_blocks = ["::/0"]
    }

    ingress_nodes_karpenter_port = {
      description                   = "Cluster API to Node group for Karpenter webhook"
      protocol                      = "tcp"
      from_port                     = 8443
      to_port                       = 8443
      type                          = "ingress"
      source_cluster_security_group = true
    }
  }

  manage_aws_auth_configmap = true
  aws_auth_roles = [
    {
      rolearn = aws_iam_role.karpenter_node.arn
      username = "system:node:{{EC2PrivateDNSName}}"
      groups = ["system:bootstrappers", "system:nodes"],
    },
  ]
}

resource "aws_iam_instance_profile" "karpenter_node" {
  name = "KarpenterNodeInstanceProfile-${local.name}"
  role = aws_iam_role.karpenter_node.name
}

resource "aws_iam_role" "karpenter_node" {
  name = "KarpenterNode-${local.name}"
  path = "/"
  assume_role_policy = jsonencode({
    Version = "2012-10-17"
    Statement = [{
      Effect = "Allow"
      Action = ["sts:AssumeRole"]
      Principal = {
	Service = "ec2.amazonaws.com"
      }
    }]
  })
  managed_policy_arns = [
    "arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy",
    "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly",
    "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
  ]
}

resource "kubectl_manifest" "karpenter_nodetemplate" {
  yaml_body = yamlencode({
    apiVersion = "karpenter.k8s.aws/v1alpha1"
    kind = "AWSNodeTemplate"
    metadata = {
      name = "default"
    }
    spec = {
      subnetSelector = {
	Name = "*private*"
	"kubernetes.io/cluster/${local.name}" = "shared"
      }
      launchTemplate = aws_launch_template.karpenter_v6prefix.name
      tags = {
	# Must match karpenter_irsa.karpenter_tag_key in generated IAM policy
	# See also https://github.com/aws/karpenter/issues/1919
	"karpenter.sh/managed-by" = module.eks.cluster_id
      }
    }
  })

  depends_on = [
    helm_release.karpenter
  ]
}

data "aws_ec2_instance_types" "karpenter" {
  # IPv6 Prefix Delegation only supported on Nitro instances
  filter {
    name = "network-info.ipv6-supported"
    values = ["true"]
  }
  filter {
    name = "hypervisor"
    values = ["nitro"]
  }
}

resource "kubectl_manifest" "karpenter_provisioner" {
  yaml_body = yamlencode({
    apiVersion = "karpenter.sh/v1alpha5"
    kind = "Provisioner"
    metadata = {
      name = "default"
    }
    spec = {
      requirements = [
	{
	  key = "node.kubernetes.io/instance-type"
	  operator = "In"
	  values = sort(data.aws_ec2_instance_types.karpenter.instance_types)
	},
	{
	  key = "karpenter.sh/capacity-type"
	  operator = "In"
	  values = ["spot"]
	},
	{
	  key = "topology.kubernetes.io/zone"
	  operator = "In"
	  values = data.aws_availability_zones.available.names
	},
      ]
      limits = {
	resources = {
	  cpu = 1000
	}
      }
      providerRef = {
	name = "default"
      }
      ttlSecondsAfterEmpty = 120
      ttlSecondsUntilExpired = 2592000
    }
  })

  depends_on = [
    helm_release.karpenter
  ]
}


data "cloudinit_config" "v6prefix" {
  gzip = false
  base64_encode = true

  part {
    content_type = "text/x-shellscript"
    # https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1958
    # Ugly because terraform doesn't export eks service cidr attribute yet.
    content = <<-USERDATA
      #!/bin/sh
      set -ex

      B64_CLUSTER_CA='${module.eks.cluster_certificate_authority_data}'
      API_SERVER_URL='${module.eks.cluster_endpoint}'
      SVC_IPV6_CIDR=$(aws eks describe-cluster --region=${local.region} --name=${local.name} --output=text --query 'cluster.{serviceIpv6Cidr: kubernetesNetworkConfig.serviceIpv6Cidr}')

      /etc/eks/bootstrap.sh '${module.eks.cluster_id}' \
      --kubelet-extra-args "--node-labels=node.k8s.aws/capacity-type=spot" \
      --b64-cluster-ca "$B64_CLUSTER_CA" \
      --apiserver-endpoint "$API_SERVER_URL" \
      --ip-family ipv6 \
      --service-ipv6-cidr "$SVC_IPV6_CIDR" \
      --use-max-pods false \
      --container-runtime containerd
    USERDATA
  }
}

data "aws_ami" "eks_optimised" {
  most_recent = true
  owners = ["amazon"]

  filter {
    name = "name"
    values = ["amazon-eks-node-${module.eks.cluster_version}-v*"]
  }
  
  filter {
    name = "root-device-type"
    values = ["ebs"]
  }

  filter {
    name = "virtualization-type"
    values = ["hvm"]
  }
}

resource "aws_launch_template" "karpenter_v6prefix" {
  name_prefix = "kv6pd-"
  description = "Karpenter node with v6 prefix attached"
  update_default_version = true

  # Boo, this means we lose arch independence
  image_id = data.aws_ami.eks_optimised.id
  
  iam_instance_profile {
    arn = aws_iam_instance_profile.karpenter_node.arn
  }

  block_device_mappings {
    device_name = "/dev/xvda"
    ebs {
      volume_size = 100
      volume_type = "gp3"
      delete_on_termination = true
    }
  }

  # user data needed for Karpenter, until it gets v6 support
  user_data = data.cloudinit_config.v6prefix.rendered
  
  monitoring {
    enabled = true
  }

  metadata_options {
    http_endpoint = "enabled"
    http_protocol_ipv6 = "disabled"
    http_put_response_hop_limit = 1
    http_tokens = "required"

    # https://github.com/aws/karpenter/issues/1527
    instance_metadata_tags = "disabled"
  }
  
  network_interfaces {
    associate_public_ip_address = false
    delete_on_termination = true

    security_groups = [module.eks.node_security_group_id]
    
    # **** Bug! ****
    #
    # Provisioning failed, launching node, creating cloud provider instance, with fleet error(s), InvalidParameterValue: You can and can only specify one of ipv6-addresses or ipv6-addresses-count  ipv6-prefix and ipv6-prefix-count.; MaxSpotInstanceCountExceeded: Max spot instance count exceeded; UnsupportedOperation: IPv6 is not supported for the instance type 'm3.medium'.
    ipv6_address_count = 1 # address for node
    ipv6_prefix_count = 1  # addresses for pods
    #
    # Comment out ipv6_prefix_count to see the simpler (but less desirable) version succeed.
    #
    # **** /Bug ****
  }

  lifecycle {
    create_before_destroy = true
  }

  tags = {
    # Must match karpenter_irsa.karpenter_tag_key in generated IAM policy
    # See also https://github.com/aws/karpenter/issues/1919
    "karpenter.sh/managed-by" = module.eks.cluster_id
  }
}

resource "aws_kms_key" "eks" {
  description = "EKS Secret Encryption Key"
  deletion_window_in_days = 7
  enable_key_rotation = true
}

data "aws_availability_zones" "available" {
  state = "available"
}

module "vpc_cni_irsa" {
  source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
  version = "~> 5.0"

  role_name = "vpc-cni-${local.name}"
  attach_vpc_cni_policy = true

  # fixme: Prefix is attached in launchtemplate instead, so don't need AssignIpv6Addresses
  vpc_cni_enable_ipv6 = true

  oidc_providers = {
    main = {
      provider_arn               = module.eks.oidc_provider_arn
      namespace_service_accounts = ["kube-system:aws-node"]
    }
  }
}

module "karpenter_irsa" {
  source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
  version = "~> 5.0"

  role_name                       = "karpenter-controller-${local.name}"
  attach_karpenter_controller_policy = true

  karpenter_tag_key               = "karpenter.sh/managed-by"
  karpenter_controller_cluster_id = module.eks.cluster_id
  karpenter_controller_node_iam_role_arns = [
    aws_iam_role.karpenter_node.arn,
  ]

  oidc_providers = {
    main = {
      provider_arn               = module.eks.oidc_provider_arn
      namespace_service_accounts = ["karpenter:karpenter"]
    }
  }
}

resource "helm_release" "karpenter" {
  namespace = "karpenter"
  create_namespace = true

  name = "karpenter"
  repository = "https://charts.karpenter.sh"
  chart = "karpenter"
  version = "v0.13.2"

  set {
    # Don't rely on coreDNS being functional
    name = "dnsPolicy"
    value = "Default"
  }
  
  set {
    name = "serviceAccount.annotations.eks\\.amazonaws\\.com/role-arn"
    value = module.karpenter_irsa.iam_role_arn
  }

  set {
    name  = "clusterName"
    value = module.eks.cluster_id
  }

  set {
    name  = "clusterEndpoint"
    value = module.eks.cluster_endpoint
  }

  set {
    name  = "aws.defaultInstanceProfile"
    value = aws_iam_instance_profile.karpenter_node.name
  }

  depends_on = [
    # Pod needs to be created after fargate profile is created :(
    module.eks.fargate_profile,
  ]
}

resource "kubectl_manifest" "testapp" {
  wait_for_rollout = false  // this is a failure demonstration

  yaml_body = yamlencode({
    apiVersion = "apps/v1"
    kind = "Deployment"
    metadata = {
      name = "testapp"
    }
    spec = {
      replicas = 1
      selector = {
	matchLabels = {app="testapp"}
      }
      template = {
	metadata = {
	  labels = {app="testapp"}
	}
	spec = {
	  containers = [{
	    name = "pause"
	    image = "public.ecr.aws/eks-distro/kubernetes/pause:3.2"
	    resources = {
	      requests = {cpu=1}
	    }
	  }]
	}
      }
    }
  })
}

module "vpc" {
  source = "terraform-aws-modules/vpc/aws"
  version = "~> 3.0"

  name = local.name
  cidr = local.cidr_v4

  azs = data.aws_availability_zones.available.names

  enable_ipv6 = true
  assign_ipv6_address_on_creation = true
  create_egress_only_igw = true

  public_subnets = [for i, v in data.aws_availability_zones.available.names : cidrsubnet(local.cidr_v4, 8, 0 + i)]
  private_subnets = [for i, v in data.aws_availability_zones.available.names : cidrsubnet(local.cidr_v4, 8, 64 + i)]
  
  public_subnet_ipv6_prefixes = [for i, v in data.aws_availability_zones.available.names : 0 + i]
  private_subnet_ipv6_prefixes = [for i, v in data.aws_availability_zones.available.names : 64 + i]
    
  enable_nat_gateway = true
  single_nat_gateway = true
  enable_dns_hostnames = true

  public_subnet_tags = {
    "kubernetes.io/cluster/${local.name}" = "shared"
    "kubernetes.io/role/elb" = 1
  }

  private_subnet_tags = {
    "kubernetes.io/cluster/${local.name}" = "shared"
    "kubernetes.io/role/internal-elb" = 1
  }
}
```
</details>

**Comment by ellistarn** [comment]
I'd love to prioritize support for IPv6. As you mention, it's possible to achieve with custom launch templates, but we need first class support for this.

**Comment by suket22** [comment]
We've got a few issues for IPV6 support floating around. Our current PR to add support https://github.com/aws/karpenter/pull/1605 is blocked on a change to be done on the eks-optimized AL2 AMI. 

I think we can close this as a duplicate of https://github.com/aws/karpenter/issues/1777

**Comment by anguslees** [comment]
> I think we can close this as a duplicate of https://github.com/aws/karpenter/issues/1777

Please don't.  This bug is not a general request for better IPv6 support, despite the re-titling.  It's a bug report about a specific combination of launchtemplate features that fails with karpenter (fleet) vs MNG (ASG).  I'm hoping the karpenter AWS devs can use their internal contacts to raise this with the fleet folks, fleet can fix this overzealous input validation bug, and we can all enjoy a better future.

**Comment by ellistarn** [comment]
Aha! Thanks for clarifying. Let me see if I can get some 👀  from @eptiger.

**Comment by ellistarn** [comment]
Shoot me an email at ellistarn@gmail.com, I don't have a DM for you anymore :)

**Comment by anguslees** [comment]
> It looks like there are a few error messages in there, and one of them is for m3.medium - I'm fairly confident we don't support ipv6 for m3.medium: https://docs.aws.amazon.com/vpc/latest/userguide/vpc-migrate-ipv6.html#vpc-migrate-ipv6-instance-types

Good call - fixed terraform file to tell karpenter to only use nitro/v6-capable instance-types.

Now (as expected) the karpenter error only mentions this ipv6-prefix-count issue:

```
ERROR	controller.provisioning	Provisioning failed, launching node, creating cloud provider instance, with fleet error(s), InvalidParameterValue: You can and can only specify one of ipv6-addresses or ipv6-addresses-count  ipv6-prefix and ipv6-prefix-count.	{"commit": "062a029"}
```

@eptiger: Thanks for investigating. Here's a recent request-id: `125db74b-6495-43e3-a4a0-13801d19ce66` (It's from a temp account set up just to investigate this issue)

**Comment by anguslees** [comment]
Ah.  It seems I hit this same error with MNG, self-managed (ASG), and a simple RunInstances - ie: there's no way to create an instance with both v6 address and prefix afaics.  I erroneously believed it worked with MNG earlier because MNG copies the provided launchtemplate, _silently ignoring the Ipv6PrefixCount value_.  Sigh.

Ok, so this issue is deeper than karpenter and fleet.  @eptiger I'm closing this as wont-fix, and I will attempt to raise it against EC2 via other channels.  Feel free to reopen/comment if tracking it here is still useful.

**Comment by eptiger** [comment]
Thanks for updating us and sorry for the confusion. I took a closer look at the code internally and misread it myself - it is, indeed, purposely failing on this case. As the error message states, you can only specify one of the following 4 parameters: ipv6-prefixes, ipv6-addresses, ipv6-prefix-count, and ipv6-address-count (I had misread the code thinking that it allowed both prefix count and address count, but it does not). 

I have opened an issue internally for our tech writers to update the public documentation to make this clearer. While I've alerted the internal team that owns this functionality of this gap, it's certainly the right thing to do for you to go through the normal means of raising a feature request to further raise visibility and explain your use case there. For the time being though, this is working as designed.
